### PR TITLE
[Shop][Account] Orders on my account page

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: app/config/config.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ imports:
     - { resource: "@SyliusCoreBundle/Resources/config/app/config.yml" }
     - { resource: "@SyliusWebBundle/Resources/config/app/config.yml" }
     - { resource: "@SyliusAdminBundle/Resources/config/app/config.yml" }
+    - { resource: "@SyliusShopBundle/Resources/config/app/config.yml" }
```
Comment (pjedrzejewski) [inline app/config/config.yml]: 👍

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php [hunk 2]
```diff
@@ -19,7 +20,18 @@
...
+     */
     public function createListQueryBuilder();
+    
+    /**
+     * @param CustomerInterface $customer
+     * @param array $sorting
+     *
+     * @return QueryBuilder
+     */
+    public function createQueryBuilderWithCustomer(CustomerInterface $customer, array $sorting = []);
```
Comment (Zales0123) [inline src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: ``createQueryBuilderByCustomer``?

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/Order/index.html.twig [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    </div>
+
+    <div class="ui stackable grid">
+        <div class="four wide column">
+            {% include 'SyliusShopBundle:Account:_main.html.twig' %}
+        </div>
+        <div class="twelve wide column">
+            <h1 class="ui monster section dividing header">
+                Order history
+                <div class="sub header">Lorem Ipsum is simply dummy text of the printing and typesetting industry.</div>
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/Order/index.html.twig]: Don't we want to have all this messages as translation keys?

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 4]
```diff
@@ -35,13 +37,18 @@
...
     private $profileUpdatePage;
-    
+
     /**
      * @var ChangePasswordPageInterface
      */
     private $changePasswordPage;
 
     /**
+     * @var IndexPage
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: ``IndexPageInterface``?

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 5]
```diff
@@ -243,6 +253,37 @@ public function iShouldBeNotifiedThatThePasswordShouldBeAtLeastCharactersLong()
...
+    {
+        $this->orderIndexPage->open();
+    }
+
+    /**
+     * @Then I should see a single order in the list
+     */
+    public function iShouldSeeASingleOrderInTheList()
+    {
+        Assert::same(
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: I would use ``Assert::eq``.

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 6]
```diff
@@ -243,6 +253,37 @@ public function iShouldBeNotifiedThatThePasswordShouldBeAtLeastCharactersLong()
...
+            1,
+            $this->orderIndexPage->countItems(),
+            '%s rows with orders should appear on page, %s rows has been found'
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: ``''%s rows with orders should appear on page, %s rows have been found.'``.

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing/account.yml [hunk 7]
```diff
@@ -1,6 +1,9 @@
 # This file is a part of the Sylius package.
 # (c) Paweł Jędrzejewski
 
+sylius_shop_account_order:
+    resource: "@SyliusShopBundle/Resources/config/routing/account/order.yml"
+
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing/account.yml]: Please import it after root and other routes.

File: src/Sylius/Bundle/ShopBundle/Resources/config/grids/account/order.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+sylius_grid:
+    grids:
+        sylius_shop_account_order:
+            driver:
+                name: doctrine/orm
+                options:
+                    class: "%sylius.model.order.class%"
+                    repository:
+                        method: createQueryBuilderWithCustomer
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/grids/account/order.yml]: createByCustomerQueryBuilder

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 9]
```diff
@@ -37,6 +37,25 @@ public function createListQueryBuilder()
...
+        $queryBuilder = $this->createQueryBuilder('o');
+
+        $queryBuilder
+            ->andWhere($queryBuilder->expr()->isNotNull('o.completedAt'))
+            ->innerJoin('o.customer', 'customer')
+            ->andWhere('customer = :customer')
+            ->setParameter('customer', $customer)
+        ;
+
+        $this->applySorting($queryBuilder, $sorting);
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: Please remove sorting, grids take care of that.

File: src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        return 'sylius_shop_order_index';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getDefinedElements()
+    {
+        return array_merge(parent::getDefinedElements(), [
+            'table' => '.table',
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php]: We should add some meaningful ids for elements. #sylius-customer-orders or something similar would make more sense than matching any table on the page.

File: src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    ) {
+        parent::__construct($session, $parameters, $router);
+
+        $this->tableAccessor = $tableAccessor;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function countItems()
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php]: ``countOrders``? It is not a generic method, so we can use something more specific here.

File: src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteName()
+    {
+        return 'sylius_shop_order_index';
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php]: Name should be something like ``sylius_shop_account_order_index``.

File: src/Sylius/Bundle/ShopBundle/Resources/config/grids/account/order.yml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+                        template: SyliusAdminBundle:Order/Field:total.html.twig
+                state:
+                    type: twig
+                    label: sylius.ui.state
+                    options:
+                        template: SyliusAdminBundle:Order/Field:state.html.twig
+            actions:
+                item:
+                    show:
+                        type: show
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/grids/account/order.yml]: I think this is out of scope for this PR.

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing/account/order.yml [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+# (c) Paweł Jędrzejewski
+
+sylius_shop_account_order:
+    resource: |
+        alias: sylius.order
+        section: shop_account
+        templates: SyliusShopBundle:Account/Order
+        except: ['create', 'update', 'delete']
+        grid: sylius_shop_account_order
+    type: sylius.resource
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing/account/order.yml]: I do not think it makes sense to generate the routing here. Let's define the routes manually - will be easier for devs to understand how to customize these in shop. (which they will do much more often than for admin stuff)

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php (hunk 2 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: @Zales0123 To me both ``createByCustomerQueryBuilder`` and ``createQueryBuilderByCustomer`` sound good, but I am leaning towards the former.

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php (hunk 5 above)
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Why? I think that `Assert::same` is more accurate and it checks if values are identical, so I guess that we should use `::same` instead of `::eq`

**Comment by pjedrzejewski** [comment]
Thanks Grzesiu! 👍

File: src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public function countOrders()
+    {
+        return $this->tableAccessor->countTableBodyRows($this->getElement('customer_orders'));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isOrderWithNumberInTheList($number)
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Page/Shop/Account/Order/IndexPage.php]: is**Single**Order
